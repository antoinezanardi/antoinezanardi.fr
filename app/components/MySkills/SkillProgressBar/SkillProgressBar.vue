<template>
  <div class="progress-container">
    <span
      id="badge-html"
      class="progress-badge"
    >
      <WrappedFontAwesomeIcon
        class="skill-icon"
        classes="me-2"
        :icon="skill.iconClasses"
        :icon-color="skill.color"
      />

      <b>
        <a
          class="skill-name text-muted"
          :href="skill.url"
          rel="noopener noreferrer"
          target="_blank"
        >
          {{ skill.name }}
        </a>
      </b>
    </span>

    <div class="progress">
      <div
        :aria-label="progressBarAriaLabel"
        aria-valuemax="100"
        aria-valuemin="0"
        class="progress-bar progress-bar-primary"
        data-aos="progress-full"
        data-aos-duration="2000"
        data-aos-offset="10"
        role="progressbar"
        :style="progressBarStyle"
      />

      <span
        class="font-weight-bold progress-value"
      >
        {{ skill.percent }}
      </span>
    </div>
  </div>
</template>

<script setup lang="ts">
import type { SkillProgressBarProps } from "~/components/MySkills/SkillProgressBar/skill-progress-bar.types";
import WrappedFontAwesomeIcon from "~/components/shared/Icons/WrappedFontAwesomeIcon/WrappedFontAwesomeIcon.vue";

const props = defineProps<SkillProgressBarProps>();

const { t } = useI18n();

const progressBarStyle = computed<Record<string, string>>(() => ({
  width: props.skill.percent,
  backgroundColor: props.skill.color,
}));

const progressBarAriaLabel = computed<string>(() => t("SkillProgressBar.skillProgress", {
  skillName: props.skill.name,
  skillPercent: props.skill.percent,
}));
</script>