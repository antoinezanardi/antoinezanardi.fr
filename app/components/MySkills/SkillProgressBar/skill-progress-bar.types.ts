import type { Skill } from "~/models/Skill";

type SkillProgressBarProps = {
  skill: Skill;
};

export type { SkillProgressBarProps };