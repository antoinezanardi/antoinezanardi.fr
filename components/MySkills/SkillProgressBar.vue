<template>
  <div class="progress-container">
    <span id="badge-html" class="progress-badge">
      <WrappedFontAwesomeIcon :icon="skill.iconClasses" :icon-color="skill.color" classes="me-2"/>
      <b>
        <a
          class="text-muted"
          :href="skill.url"
          target="_blank"
          rel="noopener noreferrer"
          v-html="skill.name"
        />
      </b>
    </span>
    <div class="progress">
      <div
        class="progress-bar progress-bar-primary"
        data-aos="progress-full"
        data-aos-offset="10"
        data-aos-duration="2000"
        role="progressbar"
        aria-valuemin="0"
        aria-valuemax="100"
        :aria-label="skill.name"
      />
      <span class="progress-value font-weight-bold" v-html="skill.percent"/>
    </div>
  </div>
</template>

<script setup lang="ts">
import WrappedFontAwesomeIcon from "~/components/shared/WrappedFontAwesomeIcon.vue";
import type { Skill } from "~/models/Skill";

defineProps<{
  skill: Skill;
}>();
</script>

<style lang="scss" scoped>
.progress {
  .progress-bar {
    width: v-bind("skill.percent");
    background-color: v-bind("skill.color");
  }
}
</style>