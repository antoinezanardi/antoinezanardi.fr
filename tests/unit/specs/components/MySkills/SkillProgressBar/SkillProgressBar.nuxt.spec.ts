import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type WrappedFontAwesomeIcon from "~/components/shared/Icons/WrappedFontAwesomeIcon/WrappedFontAwesomeIcon.vue";
import type { SkillProgressBarProps } from "~/components/MySkills/SkillProgressBar/skill-progress-bar.types";
import SkillProgressBar from "~/components/MySkills/SkillProgressBar/SkillProgressBar.vue";

describe("Skill Progress Bar Component", () => {
  const defaultProps: SkillProgressBarProps = {
    skill: {
      name: "MySkills.html",
      iconClasses: "fab fa-html5",
      color: "#E44D27",
      percent: "95%",
      url: "https://fr.wikipedia.org/wiki/HTML5",
    },
  };
  let wrapper: ReturnType<typeof mount<typeof SkillProgressBar>>;

  async function mountSkillProgressBarComponent(options: ComponentMountingOptions<typeof SkillProgressBar> = {}):
  Promise<ReturnType<typeof mount<typeof SkillProgressBar>>> {
    return mountSuspendedComponent(SkillProgressBar, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountSkillProgressBarComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Icon", () => {
    it("should pass skill icon when rendered.", () => {
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".skill-icon");

      expect(icon.props("icon")).toBe(defaultProps.skill.iconClasses);
    });

    it("should pass skill color when rendered.", () => {
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".skill-icon");

      expect(icon.props("iconColor")).toBe(defaultProps.skill.color);
    });
  });

  describe("Skill", () => {
    it("should render skill name when rendered.", () => {
      const skillName = wrapper.find<HTMLAnchorElement>(".skill-name");

      expect(skillName.text()).toBe(defaultProps.skill.name);
    });

    it("should set skill url to anchor when rendered.", () => {
      const skillUrl = wrapper.find<HTMLAnchorElement>(".skill-name");

      expect(skillUrl.attributes("href")).toBe(defaultProps.skill.url);
    });
  });

  describe("Progress Bar", () => {
    it("should set aria label to skill name when rendered.", () => {
      const progressBar = wrapper.find<HTMLDivElement>(".progress-bar");

      expect(progressBar.attributes("aria-label")).toBe(defaultProps.skill.name);
    });

    it("should set progress bar style when rendered.", () => {
      const progressBar = wrapper.find<HTMLDivElement>(".progress-bar");
      const expectedStyle = `width: ${defaultProps.skill.percent}; background-color: ${defaultProps.skill.color};`;

      expect(progressBar.attributes("style")).toBe(expectedStyle);
    });

    it("should set progress bar percent in progress bar label when rendered.", () => {
      const progressBarLabel = wrapper.find<HTMLDivElement>(".progress-value");

      expect(progressBarLabel.text()).toBe(defaultProps.skill.percent);
    });
  });
});