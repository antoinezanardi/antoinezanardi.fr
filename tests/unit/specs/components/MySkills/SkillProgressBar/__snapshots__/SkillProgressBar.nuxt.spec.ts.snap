// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skill Progress Bar Component > should match snapshot when rendered. 1`] = `
"<div class="progress-container"><span id="badge-html" class="progress-badge"><wrapped-font-awesome-icon-stub icon="fab fa-html5" iconcolor="#E44D27" classes="me-2" class="skill-icon"></wrapped-font-awesome-icon-stub><b><a class="skill-name text-muted" href="https://fr.wikipedia.org/wiki/HTML5" rel="noopener noreferrer" target="_blank">MySkills.html</a></b></span>
  <div class="progress">
    <div aria-label="SkillProgressBar.skillProgress, {&quot;skillName&quot;:&quot;MySkills.html&quot;,&quot;skillPercent&quot;:&quot;95%&quot;}" aria-valuemax="100" aria-valuemin="0" class="progress-bar progress-bar-primary" data-aos="progress-full" data-aos-duration="2000" data-aos-offset="10" role="progressbar" style="width: 95%; background-color: #E44D27;"></div><span class="font-weight-bold progress-value">95%</span>
  </div>
</div>"
`;
